### PR TITLE
Fix bugs related to uninitialised GPU memory and istep

### LINKED ISSUE #2720: istep not passed to loss modules
# Description

istep is being passed from trainer.py to loss_calculator.py, but from here it doesn't get passed to the loss modules.
This is a bug, e.g. when using context loss warm_up in the JEPA case as the istep will forever be zero. Even worse it
is silent!

### PR DESCRIPTION
**Pull request by sophie-xhonneux**
## Description

Change torch.empty to torch.zeros in StreamEmbedder to ensure there is no uninitialised GPU memory, also skip streams
with stream embed Identity when considering memory to allocate (#2721)

Fix istep not being passed along in loss_calculator.py (#2720)

Config changes for space JEPA (no need to review)

Remark: Linting is broken, but this was true before.

## Issue Number

Closes #2720 #2721

Is this PR a draft? Mark it as draft.

## Checklist before asking for review

-   [x] I have performed a self-review of my code
-   [ ] My changes comply with basic sanity checks:
      - I have fixed formatting issues with `./scripts/actions.sh lint`
      - I have run unit tests with `./scripts/actions.sh unit-test`
      - I have documented my code and I have updated the docstrings.
      - I have added unit tests, if relevant
-   [ ] I have tried my changes with data and code:
      - I have run the integration tests with `./scripts/actions.sh integration-test`
      - (bigger changes) I have run a full training and I have written in the comment the run_id(s): `launch-slurm.py --time 60`
      - (bigger changes and experiments) I have shared a hegdedoc in the github issue with all the configurations and runs for this experiments
-   [x] I have informed and aligned with people impacted by my change:
    - for config changes: the MatterMost channels and/or a design doc
    - for changes of dependencies: the MatterMost software development channel

### CODE REVIEW

File: src/weathergen/model/engines.py [hunk 1]
```diff
@@ -1171,14 +1181,19 @@ class DeepSSLFusion(nn.Module):
 
     def __init__(self, num_levels: int, dim_embed: int, hidden_factor: int = 2):
         super().__init__()
+        # Intermediate encoder levels are tapped before the encoder's final LayerNorm, so they
+        # arrive on very different (and much larger) scales than the final level. Normalize each
+        # level before concatenating so the fusion and the downstream SSL head see a common scale.
+        self.level_norms = nn.ModuleList([nn.LayerNorm(dim_embed) for _ in range(num_levels)])
```
Comment (clessig) [inline src/weathergen/model/engines.py:1187]: This change seems unrelated to the problem that is addressed in the PR?

File: src/weathergen/model/engines.py [hunk 2]
```diff
@@ -1171,14 +1181,19 @@ class DeepSSLFusion(nn.Module):
...
         self.proj = nn.Sequential(
             nn.Linear(num_levels * dim_embed, hidden_factor * dim_embed, bias=False),
             nn.GELU(),
             nn.Linear(hidden_factor * dim_embed, dim_embed, bias=False),
         )
 
     def forward(self, levels: list[torch.Tensor]) -> torch.Tensor:
-        return self.proj(torch.cat(levels, dim=-1))
+        normed = [norm(level) for norm, level in zip(self.level_norms, levels, strict=True)]
```
Comment (clessig) [inline src/weathergen/model/engines.py:1195]: Same as above

File: src/weathergen/model/model.py [hunk 3]
```diff
@@ -690,10 +690,10 @@ def print_num_parameters(self) -> None:
             np.prod(self.encoder.q_cells.shape) if self.encoder.q_cells.requires_grad else 0
         )
 
-        if self.encoder.q_aux:
-            num_params_q_aux = (
-                np.prod(self.encoder.q_aux.shape) if self.encoder.q_aux.requires_grad else 0
-            )
+        # if self.encoder.q_aux:
```
Comment (clessig) [inline src/weathergen/model/model.py]: This change was introduced to make FSDP work

File: src/weathergen/model/model.py [hunk 4]
```diff
@@ -736,8 +736,8 @@ def print_num_parameters(self) -> None:
         print(f" Local assimilation engine: {num_params_ae_local:,}")
         print(f" Local-global adapter: {num_params_ae_adapter:,}")
         print(f" Learnable spatial queries: {num_params_q_cells:,}")
-        if self.encoder.q_aux:
-            print(f" Learnable auxiliary queries: {num_params_q_aux:,}")
+        # if self.encoder.q_aux:
```
Comment (clessig) [inline src/weathergen/model/model.py]: Same as above

**Review by clessig — COMMENTED** [review]
Minor comments. This PR seems to have some unrelated changes and would break FSDP again, I think.

File: config/streams/era5_georing_avhrr_synop_lowres_sjepa/avhrr.yml [hunk 5]
```diff
@@ -0,0 +1,86 @@
+# (C) Copyright 2024 WeatherGenerator contributors.
```
Comment (shmh40) [inline config/streams/era5_georing_avhrr_synop_lowres_sjepa/avhrr.yml:1]: Configs

File: src/weathergen/model/model.py (hunk 3 above)
Comment (shmh40) [inline src/weathergen/model/model.py]: Need to revert this I think

File: src/weathergen/train/loss_modules/loss_module_ssl.py [hunk 6]
```diff
@@ -321,7 +321,10 @@ def _masked_l1_loss_per_sample(
     mask_f = mask.to(per_token_loss.dtype)
     token_counts = mask_f.sum(dim=-1)
     valid_samples = token_counts > 0
-    per_sample_loss = (per_token_loss * mask_f).sum(dim=-1) / token_counts.clamp(min=1.0)
+    # torch.where (not multiply): NaN at mask-excluded positions must not poison the sum.
```
Comment (shmh40) [inline src/weathergen/train/loss_modules/loss_module_ssl.py:324]: Can merge this PR in last so it is after the PR that explicitly does this

File: src/weathergen/model/model.py (hunk 3 above)
Comment (sophie-xhonneux) [reply src/weathergen/model/model.py]: this broke in single GPU mode for me :/ can you explain the error to me?

Comment (sophie-xhonneux) [reply src/weathergen/model/model.py]: see my comment below, the statement breaks in single GPU mode if q_aux exists because it is a nn.Parameter tensor so not bool or can be reduced to bool

Comment (shmh40) [reply src/weathergen/model/model.py]: Here is the PR:

https://github.com/ecmwf/WeatherGenerator/pull/2686

Comment (sophie-xhonneux) [reply src/weathergen/model/model.py]: we are in the function that prints parameters, does not seem like it will break FSDP

File: src/weathergen/model/model.py (hunk 4 above)
Comment (sophie-xhonneux) [reply src/weathergen/model/model.py]: we are printing parameters, this should not break FSDP, it doesn't create the parameters

File: src/weathergen/model/model.py (hunk 3 above)
Comment (shmh40) [reply src/weathergen/model/model.py]: Yes I think you're right it is not fsdp but whether it breaks I think depends if you have register tokens or not right? 

You do have register tokens (I guess), so now it works. But if we want to run with 0 register tokens with this code, we are now doing shape of None and we get an error?

Comment (shmh40) [reply src/weathergen/model/model.py]: So I think we should actually have: 

`
if self.encoder.q_aux is not None:
`

Comment (sophie-xhonneux) [reply src/weathergen/model/model.py]: yes! you are right, I couldn't think of the right check at midnight yesterday so I commented it out, but what you right is correct I think

File: packages/common/src/weathergen/common/config.py [hunk 7]
```diff
@@ -140,7 +140,7 @@ def _strip_interpolation(conf: Config) -> Config:
     stripped = {}
     if OmegaConf.is_dict(conf):
         for key in list(conf.keys()):
-            key = str(key)
+            # key = str(key) # this lines breaks target_source_correspondence please stop re introducing it
```
Comment (shmh40) [inline packages/common/src/weathergen/common/config.py]: Do we need this for anything else in the config?

I get around it with e.g. `target_source_correspondence: {"0": {"0": independent}}` in target source I think, in case it is necessary for something else in the config

Comment (sophie-xhonneux) [reply packages/common/src/weathergen/common/config.py]: I would ask the question the other way around, is there any reason this line is needed, it breaks the type assumptions of YAML

**Review by clessig — COMMENTED** [review]
Can we clean up the PR so that we don't change 20 files unrelated to the problems. In engines.py are also changes that are unrelated?

File: config/streams/era5_georing_avhrr_synop_lowres_sjepa/avhrr.yml (hunk 5 above)
Comment (clessig) [reply config/streams/era5_georing_avhrr_synop_lowres_sjepa/avhrr.yml:1]: They should go to a separate repository

**Review by clessig — APPROVED** [review]
